src/overlays/actors/ovl_Bg_Mori_Hashira4/z_bg_mori_hashira4.c

```diff
@@ -20,27 +27,139 @@ const ActorInit Bg_Mori_Hashira4_InitVars = {
+                     this->dyna.actor.id, this->dyna.actor.params);
+    }
+}
+
+void BgMoriHashira4_Init(Actor* thisx, GlobalContext* globalCtx) {
+    BgMoriHashira4* this = THIS;
+
+    this->switchFlag = (this->dyna.actor.params >> 8) & 0x3F;
+    this->dyna.actor.params &= 0xFF;
+
+    if (this->dyna.actor.params == 0) {
+        BgMoriHashira4_InitDynaPoly(this, globalCtx, &D_06001AF8, DPM_UNK3);
+    } else {
+        BgMoriHashira4_InitDynaPoly(this, globalCtx, &D_060089E0, DPM_UNK);
+    }
+    Actor_ProcessInitChain(&this->dyna.actor, sInitChain);
+    this->objBankIndex = Object_GetIndex(&globalCtx->objectCtx, 0x73);
+    if (this->objBankIndex < 0) {
+        Actor_Kill(&this->dyna.actor);
+        osSyncPrintf("Error : バンク危険！(arg_data 0x%04x)(%s %d)\n", this->dyna.actor.params,
+                     "../z_bg_mori_hashira4.c", 0xC4);
+    } else if ((this->dyna.actor.params != 0) && Flags_GetSwitch(globalCtx, this->switchFlag)) {
+        Actor_Kill(&this->dyna.actor);
+    } else {
+        Actor_SetHeight(&this->dyna.actor, 50.0f);
+        BgMoriHashira4_SetupObjectCheck(this);
+        osSyncPrintf("(森の神殿 ４本柱)(arg_data 0x%04x)\n", this->dyna.actor.params);
+        unkTimer = 0;
+    }
+}
+
+void BgMoriHashira4_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    BgMoriHashira4* this = THIS;
+
+    DynaPolyInfo_Free(globalCtx, &globalCtx->colCtx.dyna, this->dyna.dynaPolyId);
+}
+
+void BgMoriHashira4_SetupObjectCheck(BgMoriHashira4* this) {
+    BgMoriHashira4_SetActionFunc(this, BgMoriHashira4_ObjectCheck);
+}
+
+void BgMoriHashira4_ObjectCheck(BgMoriHashira4* this, GlobalContext* globalCtx) {
+    if (Object_IsLoaded(&globalCtx->objectCtx, this->objBankIndex)) {
+        this->gateTimer = 0;
+        if (this->dyna.actor.params == 0) {
+            BgMoriHashira4_SetupPillars(this);
+        } else {
+            BgMoriHashira4_SetActionFunc(this, BgMoriHashira4_GateWait);
+        }
+        this->dyna.actor.draw = BgMoriHashira4_Draw;
+    }
+}
+
+void BgMoriHashira4_SetupPillars(BgMoriHashira4* this) {
+    BgMoriHashira4_SetActionFunc(this, BgMoriHashira4_PillarsRotate);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Mori_Hashira4/BgMoriHashira4_Init.s")
+void BgMoriHashira4_PillarsRotate(BgMoriHashira4* this, GlobalContext* globalCtx) {
+    this->dyna.actor.shape.rot.y = this->dyna.actor.posRot.rot.y += 0x96;
+    Audio_PlayActorSound2(&this->dyna.actor, NA_SE_EV_ROLL_STAND_2 - SFX_FLAG);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Mori_Hashira4/BgMoriHashira4_Destroy.s")
+void BgMoriHashira4_GateWait(BgMoriHashira4* this, GlobalContext* globalCtx) {
+    if (Flags_GetSwitch(globalCtx, this->switchFlag) || (this->gateTimer != 0)) {
+        this->gateTimer++;
+        if (this->gateTimer > 30) {
+            Audio_PlayActorSound2(&this->dyna.actor, NA_SE_EV_METALDOOR_OPEN);
+            BgMoriHashira4_SetActionFunc(this, &BgMoriHashira4_GateOpen);
```

Comment:
the single `&ActionFunc` (`&BgMoriHashira4_GateOpen`), we were this close to perfection